### ISSUE TITLE
Concurrent pings hide missing device

### ISSUE THREAD

**Issue by codeedog** [quantumsheep/node-icmp#5]
If more than one ping is issued at the same time and one of the ping'ed IP addresses should fail (no device at that IP), the code incorrectly responds with a successful ping. I looked through node-icmp and raw-socket. I suspect there's a data structure reuse issue in raw-socket (or some sort of failure to copy local context), but I don't know enough about the use and workings of Nan::Call to analyze and debug.

I've attached a sample js program with comments that include output demonstrating the problem.

Let me know if you have any questions.

[bug.txt](https://github.com/quantumsheep/node-icmp/files/4633965/bug.txt)

EDIT: Including output here:

````
> ping -c 3 -t 3 192.168.1.33
PING 192.168.1.33 (192.168.1.33): 56 data bytes
Request timeout for icmp_seq 0
Request timeout for icmp_seq 1

--- 192.168.1.33 ping statistics ---
3 packets transmitted, 0 packets received, 100.0% packet loss

> node bug.js correct
192.168.1.33: 192.168.1.33, FAILED

> node bug.js bug
192.168.1.33: 192.168.1.33, 13.106855000000001 ms
8.8.8.8: 8.8.8.8, 13.81142 ms

> sw_vers
ProductName:	Mac OS X
ProductVersion:	10.13.6
BuildVersion:	17G10021
````

**Comment by quantumsheep** [comment]
Thanks you for sharing your debug code! I was able to successfully replicate the issue.

So, it's seems to happen "sometimes", it may effectively be a data leak from a dependency like `raw-socket` or `nan` but I dug into the code and found nothing, even `nan` seems to be pretty safe and not using a lot of pointers that can be sharing/leaking data.

I though it might come from `node-icmp` itself but it creates a new socket every time the static function `ping` is called so every data is theoretically separated.

It could also come from the Node's event loop overflowing the network output but i doubt it's the case.

Although I'm also on a Mac (running Mac OS X 10.15.4) so it could be an issue related to the operating system or the hardware from our computers.

_tl;dr: idk, help is needed._

**Comment by codeedog** [comment]
Thanks for the feedback. I'm glad you were able to replicate it. If I have some time, I'll try to reproduce on linux or windows.

**Comment by angristan** [comment]
FYI I was able to reproduce it on Linux

**Comment by quantumsheep** [comment]
I finally figured this out! The `message` event is emitting every ICMP responses.

That might be due to ICMP packets being not identifiable and thus can't be linked to a specific network request.

https://github.com/quantumsheep/node-icmp/blob/b2177688b0813928feb9d77d6740e37ff7c2b15b/index.js#L70

My fix for this was to check if the source IP of the packet is the one we requested. That might not be the best solution because it could concurrence other ICMP calls to the same IP but that fixes the current issue.

https://github.com/quantumsheep/node-icmp/blob/1422642701f7b0c9b0b1a57d2d87091563dd9515/index.js#L70-L71
